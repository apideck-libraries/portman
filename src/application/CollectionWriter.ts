import { CollectionDefinition } from 'postman-collection'
import {
  injectEnvVariables,
  orderCollectionRequests,
  overwriteCollectionKeyValues,
  overwriteCollectionValues,
  overwriteCollectionSecurityValues,
  writeCollectionPreRequestScripts,
  writeRawReplacements
} from '.'
import { GlobalConfig, PortmanConfig, PortmanOptions } from '../types'
import { isEmptyObject } from 'utils'

export class CollectionWriter {
  public collection: CollectionDefinition
  config: PortmanConfig
  options: PortmanOptions

  constructor(config: PortmanConfig, options: PortmanOptions, collection: CollectionDefinition) {
    this.config = config
    this.options = options
    this.collection = collection
  }

  public execute(): void {
    const { envFile, baseUrl } = this.options

    const { globals = {} } = this.config
    const {
      collectionPreRequestScripts = [],
      securityOverwrites = {},
      keyValueReplacements = {},
      valueReplacements = {},
      rawReplacements = [],
      orderOfOperations = []
    } = globals as GlobalConfig

    let collection = this.collection

    // --- Portman - Set Security values for Postman
    if (securityOverwrites && !isEmptyObject(securityOverwrites)) {
      collection = overwriteCollectionSecurityValues(collection, securityOverwrites)
    }

    // --- Portman - Search for keys in dictionary to set the value if key is found anywhere in collection
    if (keyValueReplacements && !isEmptyObject(keyValueReplacements)) {
      collection = overwriteCollectionKeyValues(collection, keyValueReplacements)
    }

    // --- Portman - Search for keys in dictionary to set the values if value is found anywhere in collection
    if (valueReplacements && !isEmptyObject(valueReplacements)) {
      collection = overwriteCollectionValues(collection, valueReplacements)
    }

    // --- Portman - Use .env to inject variables on Collection
    if (envFile) {
      collection = injectEnvVariables(collection, envFile, baseUrl)
    }

    // --- Portman - Set manually order Postman requests
    if (orderOfOperations && orderOfOperations.length > 0) {
      collection = orderCollectionRequests(collection, orderOfOperations)
    }

    // --- Portman - Set Postman pre-requests
    if (collectionPreRequestScripts && collectionPreRequestScripts.length > 0) {
      collection = writeCollectionPreRequestScripts(collection, collectionPreRequestScripts)
    }

    // --- Portman - Replace & clean-up Postman
    if (rawReplacements && rawReplacements.length > 0) {
      const collectionString = writeRawReplacements(
        JSON.stringify(collection, null, 2),
        rawReplacements
      )
      this.collection = JSON.parse(collectionString)
    } else {
      this.collection = collection
    }
  }
}
