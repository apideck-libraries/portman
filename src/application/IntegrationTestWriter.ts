import { Collection, Item, ItemGroup } from 'postman-collection'
import { IntegrationTestConfig, VariationConfig, PortmanTestTypes } from '../types'
import { TestSuite, VariationWriter } from './'

export class IntegrationTestWriter {
  public testSuite: TestSuite
  integrationTestFolder: ItemGroup<Item>
  integrationTestCollection: Collection

  constructor({
    testSuite,
    integrationTestFolderName
  }: {
    testSuite: TestSuite
    integrationTestFolderName: string
  }) {
    this.testSuite = testSuite
    this.integrationTestFolder = new ItemGroup<Item>({
      name: integrationTestFolderName
    })
    this.integrationTestCollection = new Collection()
  }

  public add(integrationTest: IntegrationTestConfig): void {
    const { testSuite } = this
    const { name, operations } = integrationTest

    const variationWriter = new VariationWriter({
      testSuite: this.testSuite,
      variationFolderName: name
    })

    operations.map(operation => {
      const variations = (operation?.variations || []) as VariationConfig[]
      const { openApiOperationId } = operation

      const pmOperation = testSuite.postmanParser.getOperationById(openApiOperationId)

      if (!pmOperation) return

      const oaOperation = testSuite.oasParser.getOperationByPath(pmOperation.pathRef)

      // Handle integration variations
      variations.map(variation => {
        const variationName = variation.name

        const operationVariation = pmOperation.clone({
          newId: `${variationName}-${Math.random()}`,
          name: variationName
        })

        // Set/Update Portman operation test type
        this.testSuite.registerOperationTestType(operationVariation, PortmanTestTypes.integration)

        variationWriter.injectVariations(operationVariation, oaOperation, variation, operation)
        variationWriter.addToFolder(operationVariation, variationWriter.variationFolder)
      })
    })

    this.integrationTestCollection = variationWriter.mergeToCollection(
      this.integrationTestCollection
    )
  }

  public mergeToCollection(collection: Collection): Collection {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    this.integrationTestCollection.items.map((item: any) => {
      this.integrationTestFolder.items.add(item)
    })

    collection.items.add(this.integrationTestFolder)
    return collection
  }
}
