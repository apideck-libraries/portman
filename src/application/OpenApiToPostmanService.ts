import fs from 'fs-extra'
import oaConverter from 'openapi-to-postmanv2'
import ora from 'ora'
import path from 'path'

export class OpenApiToPostmanService {
  /**
   * Helper function for the CLI to convert OpenApi data input
   * (ported from https://github.com/postmanlabs/openapi-to-postman)
   * @param {String} openApiObj - OpenApi spec file used for conversion input
   * @param options - OpenApi to Postman conversion options
   * @returns {string}
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
  async convert(openApiObj: string, options: any): Promise<string> {
    return await new Promise((resolve, reject) => {
      const inputFile = path.resolve(options.inputFile)
      openApiObj = fs.readFileSync(inputFile, 'utf8')

      // apply options from config file if present
      if (options.configFile) {
        const configFile = path.resolve(options.configFile)
        // console.log('Options Config file: ', configFile)
        const configFileOptions = JSON.parse(fs.readFileSync(configFile, 'utf8'))
        options = Object.assign({}, options, configFileOptions)
      }

      if (options.testsuiteFile) {
        const testsuiteFile = path.resolve(options.testsuiteFile)
        // console.log('Testsuite file: ', testsuiteFile)
        options.testSuite = true
        options.testSuiteSettings = JSON.parse(fs.readFileSync(testsuiteFile, 'utf8'))
      }

      const spinner = ora({
        prefixText: ' ',
        text: 'Converting OpenApi to Postman Collection'
      }).start()

      // Convert OpenApi to Postman collection
      oaConverter.convert(
        {
          type: 'string',
          data: openApiObj
        },
        options,
        (err, status) => {
          if (err) {
            spinner.fail(err.toString())
            reject(err)
          }
          if (!status.result) {
            // console.log(status.reason)
            spinner.fail(status.reason)
            reject(status.reason)
          } else if (options.outputFile) {
            const filePath = path.resolve(options.outputFile)

            try {
              fs.writeFileSync(filePath, JSON.stringify(status.output[0].data, null, 4))
            } catch (error) {
              console.log('Could not write to file', error)
              spinner.fail(error.toString())
              reject(error)
            }

            // Return Postman collection
            spinner.succeed('Conversion successful')
            resolve(status.output[0].data)
          } else {
            spinner.succeed('Conversion successful')
            resolve(status.output[0].data)
          }
        }
      )
    })
  }
}
