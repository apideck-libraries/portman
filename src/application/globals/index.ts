export * from './injectEnvVariables'
export * from './orderCollectionRequests'
export * from './overwriteCollectionKeyValues'
export * from './overwriteCollectionSecurityValues'
export * from './overwriteCollectionValues'
export * from './renamePostmanCollection'
export * from './stripResponseExamples'
export * from './writeCollectionPreRequestScripts'
export * from './writeRawReplacements'
