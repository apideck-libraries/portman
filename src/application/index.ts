export * from './CollectionWriter'
export * from './globals'
export * from './newman'
export * from './overwrites'
export * from './tests'
export * from './TestSuite'
export * from './variables'
