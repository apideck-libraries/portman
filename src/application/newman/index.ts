export * from './runNewmanWith'
export * from './writeNewmanEnv'
