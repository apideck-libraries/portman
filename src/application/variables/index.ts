export * from './assignVarFromRequestBody'
export * from './assignVarFromResponseBody'
export * from './assignVarFromResponseHeader'
export * from './assignVarFromValue'
