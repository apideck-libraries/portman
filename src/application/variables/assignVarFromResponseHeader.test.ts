import { getPostmanMappedOperation } from '../../../__tests__/testUtils/getPostmanMappedOperation'
import { assignVarFromResponseHeader } from '../../application'
import { PostmanMappedOperation } from '../../postman'
import { GlobalConfig } from '../../types'

describe('assignVarFromResponseHeader', () => {
  let pmOperation: PostmanMappedOperation

  beforeEach(async () => {
    pmOperation = await getPostmanMappedOperation()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should add postman collection var with name and response header value', async () => {
    const varSetting = {
      responseHeaderProp: 'operation-location',
      name: 'leadsAdd.header'
    }

    pmOperation = assignVarFromResponseHeader(varSetting, pmOperation)
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should show the log output for add postman collection var with name and response header value', async () => {
    const varSetting = {
      responseHeaderProp: 'operation-location',
      name: 'leadsAdd.header'
    }

    pmOperation = assignVarFromResponseHeader(varSetting, pmOperation, {
      logAssignVariables: true
    })
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should not show the log output for add postman collection var with name and response header value', async () => {
    const varSetting = {
      responseHeaderProp: 'operation-location',
      name: 'leadsAdd.header'
    }

    pmOperation = assignVarFromResponseHeader(varSetting, pmOperation, {
      logAssignVariables: false
    })
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should generate a variable for the response header and convert the casing for the variable name', async () => {
    const varSetting = {
      responseHeaderProp: 'portman'
    }
    const global = { variableCasing: 'snakeCase' } as GlobalConfig
    pmOperation = assignVarFromResponseHeader(varSetting, pmOperation, {}, global)
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should generate a variable for the response header and not convert the casing for the variable name', async () => {
    const varSetting = {
      responseHeaderProp: 'portman'
    }
    const global = {} as GlobalConfig
    pmOperation = assignVarFromResponseHeader(varSetting, pmOperation, {}, global)
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })
})
