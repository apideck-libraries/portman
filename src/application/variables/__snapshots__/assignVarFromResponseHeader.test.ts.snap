// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignVarFromResponseHeader should add postman collection var with name and response header value 1`] = `
Array [
  "// Set response object as internal variable
let jsonData = {};
try {jsonData = pm.response.json();}catch(e){}
",
  "// pm.collectionVariables - Set leadsAdd.header as variable for header 
let leadsAddheader = pm.response.headers.get(\\"operation-location\\");
if (leadsAddheader !== undefined) {
   pm.collectionVariables.set(\\"leadsAdd.header\\", leadsAddheader);
   console.log(\\"- use {{leadsAdd.header}} as collection variable for value\\", leadsAddheader);
};
",
]
`;

exports[`assignVarFromResponseHeader should generate a variable for the response header and convert the casing for the variable name 1`] = `
Array [
  "// Set response object as internal variable
let jsonData = {};
try {jsonData = pm.response.json();}catch(e){}
",
  "// pm.collectionVariables - Set companies_one_portman as variable for header 
let companiesoneportman = pm.response.headers.get(\\"portman\\");
if (companiesoneportman !== undefined) {
   pm.collectionVariables.set(\\"companies_one_portman\\", companiesoneportman);
   console.log(\\"- use {{companies_one_portman}} as collection variable for value\\", companiesoneportman);
};
",
]
`;

exports[`assignVarFromResponseHeader should generate a variable for the response header and not convert the casing for the variable name 1`] = `
Array [
  "// Set response object as internal variable
let jsonData = {};
try {jsonData = pm.response.json();}catch(e){}
",
  "// pm.collectionVariables - Set companiesOne.portman as variable for header 
let companiesOneportman = pm.response.headers.get(\\"portman\\");
if (companiesOneportman !== undefined) {
   pm.collectionVariables.set(\\"companiesOne.portman\\", companiesOneportman);
   console.log(\\"- use {{companiesOne.portman}} as collection variable for value\\", companiesOneportman);
};
",
]
`;

exports[`assignVarFromResponseHeader should not show the log output for add postman collection var with name and response header value 1`] = `
Array [
  "// Set response object as internal variable
let jsonData = {};
try {jsonData = pm.response.json();}catch(e){}
",
  "// pm.collectionVariables - Set leadsAdd.header as variable for header 
let leadsAddheader = pm.response.headers.get(\\"operation-location\\");
if (leadsAddheader !== undefined) {
   pm.collectionVariables.set(\\"leadsAdd.header\\", leadsAddheader);
   // console.log(\\"- use {{leadsAdd.header}} as collection variable for value\\", leadsAddheader);
};
",
]
`;

exports[`assignVarFromResponseHeader should show the log output for add postman collection var with name and response header value 1`] = `
Array [
  "// Set response object as internal variable
let jsonData = {};
try {jsonData = pm.response.json();}catch(e){}
",
  "// pm.collectionVariables - Set leadsAdd.header as variable for header 
let leadsAddheader = pm.response.headers.get(\\"operation-location\\");
if (leadsAddheader !== undefined) {
   pm.collectionVariables.set(\\"leadsAdd.header\\", leadsAddheader);
   console.log(\\"- use {{leadsAdd.header}} as collection variable for value\\", leadsAddheader);
};
",
]
`;
